### ISSUE TITLE
Request for a simple clipboard manager

### ISSUE THREAD

**Issue by bonbonsiregar** [rkkr/simple-keyboard#338]
There's a [keyboard app](https://play.google.com/store/apps/details?id=com.simplemobiletools.keyboard) in Play Store with the same name but with added clipboard manager, I think that would be a good feature to add to this app.

**Comment by JM480** [comment]
Do you mean like #95 or is this different from that feature?

**Comment by bonbonsiregar** [comment]
> Do you mean like #95 or is this different from that feature?

more like the clipboard manager that gboard have

**Comment by smaragdus** [comment]
@rkkr

First, I would like to thank the developer for this excellent keyboard!

What prevents me from using [Simple Keyboard](https://github.com/rkkr/simple-keyboard) as my default keyboard is the lack of clipboard manager as I do a lot of copy/paste.

I would appreciate a clipboard manager like the one available in [FlorisBoard](https://github.com/florisboard/florisboard).

A clipboard manager would be extremely useful for those users who copy a lot especially having in mind that since Android 10 background clipboard access has been limited and external clipboard managers no longer work.

**Comment by rkkr** [comment]
FlorisBoard and all other keyboards utilize the "suggestions bar" above the keyboard for selection of clipboard text. It's not available in SimpleKeyboad. I'm not against clipboard support, there's just no good place to add it. And I won't add a top bar just for clipboard.

**Comment by smaragdus** [comment]
@rkkr 

First, thank you for the quick response.

> FlorisBoard and all other keyboards utilize the "suggestions bar" above the keyboard for selection of clipboard text. It's not available in SimpleKeyboad.

I don't use 'suggestions bar', I have disabled it and I have even forgotten it existed. Version 0.3.16 (86) (F-Droid edition) looks like this on my phone:

![Screenshot_20220925-173653](https://user-images.githubusercontent.com/9025757/192154665-b9651e90-f7a0-4726-b420-d706ae13498b.png)

There are 3 clipboard buttons on the upper row of the tool bar, from right to left:

1. show clipboard history;
2. delete current clipboard item;
3. paste current clipboard item;

I find button 2 useless, the other two buttons, 1 & 3, **show clipboard history** and **paste paste current clipboard item** are the ones that matter. Clipboard history looks like this:

![Screenshot_20220925-173706](https://user-images.githubusercontent.com/9025757/192155169-9ff7dc70-719e-41d3-b1b3-2791d2646ef2.png)

The user can define in settings how many clipboard items should be kept.

My idea- you may add a tool bar which may be disabled by default which may contain some buttons for useful commands like:

- paste clipboard;
- show clipboard history;
- select all;
- copy;
- cut;
- undo;
- redo;

Such a tool bar which can be enabled in settings like the number row would be of great usefulness to me. What do you think?

Right now I don't have access to my computer so I could not produce better (edited) screens.

Regards

**Comment by rkkr** [comment]
Technically it's the same suggestion bar, just showing buttons instead of words. The two rows of buttons above the keyboard is exactly the horror I don't want to add. You need the clipboard 3 times a day and you have that thing taking screen space for the rest of the time.

As for number row- that is available.

**Comment by smaragdus** [comment]
@rkkr 

A clarification (perhaps unneeded)- the upper row of the tool bar can be hidden (the last button of the lower row to the right shows/hides the upper row of the tool bar):

**two rows tool bar**

![Screenshot_20220926-123004](https://user-images.githubusercontent.com/9025757/192244301-ef8f3aaf-e743-45ba-a5ed-d0b41a0d9d1a.png)

**one row tool bar**

![Screenshot_20220926-123027](https://user-images.githubusercontent.com/9025757/192244664-c118cf72-ad52-4561-bf95-7c445b525e4d.png)

The grey tool bar is part of the text editor, not of the keyboard.

One of my problems with FlorisBoard is that the tool bar is not customizable- the uset can not re-arrange the buttons (if I could I would have removed the emoji button for example).

I think that I understand your design decision to keep your keyboard as light and as minimal as possible. My idea was that the additional, single-row tool bar where the clipboard manager is placed would be disabled by default (just like the number row) and enabled by those users who prefer to use the clipboard manager.

I suppose that a clipboard mannager cannot be added to Simple Keyboard without the implentation of a tool bar from where the clipboard mannager is accesseible.

I use clipboard managers on Windows a lot and one of the primary reasons to start searching for open source keyboards was the very poor clipboard mannager in the horrible monstrosity Gboard (which is even worse than the SwiftKey abomination).

I am sorry I was not able to express my opinion in a more concise manner.

**Comment by rkkr** [comment]
Your opinion is not wrong :) I'm also not against a clipboard support, I just never figured a way to add it without a permanent top bar (which takes screen space), or permanent show/hide button (which takes keys space).

**Comment by licaon-kter** [comment]
Long press BACKSPACE? SPACE?

**Comment by lianergoist** [comment]
> I'm also not against a clipboard support, I just never figured a way to add it without a permanent top bar (which takes screen space), or permanent show/hide button (which takes keys space).

I have used a Samsung tablet for years, and I must admit I like the keyboard. It use more space, but it has a few nice features.

1. delete key
2. arrows move cursor left and right
3. Ctrl key

I agree, we want to keep things simple and clean, so maybe Simple Keyboard don't need delete and arrows. But the Ctrl key does allow keyboard shortcuts - Ctrl-c, Ctrl-x, Ctrl-v, Ctrl-a, etc., and I think there are room for such a key. There are some free space to the right of the space-bar - and it takes up a lot of space. Does it really have to be that wide? I think it should be possible to add a Ctrl key...

![Screenshot_20221021-092357](https://user-images.githubusercontent.com/6023573/197142253-c2162394-779c-423c-8174-bf7b2f55476f.jpg)

**Comment by rkkr** [comment]
Tablet layouts is a separate story. There is space for arrows and ctrl and many other things, even GBoard now has extended the tablet layout. However that is just 5.7% of devices and for 94.3% the the problem isn't solved.

**Comment by lianergoist** [comment]
> Tablet layouts is a separate story. 

You are right. I have now tried to install Simple Keyboard on my phone, and yes, there are no space for new keys. You mentioned the comma key. I think this could be a good solution. Maybe it could toggle between Ctrl and comma? You could write a fancy clipboard manager, like suggested, but I think a Ctrl key could be more powerful. Do what you like. And thanks for making this beautiful keyboard!

**Comment by smaragdus** [comment]
My opinion is that it would be extremely useful if there is an option to enable an additional top bar in settings (which may be disabled by default) which contains additional commands (select all, copy, cut, undo, redo, paste, clipboard history, etc) which will add both flexibility and choice- those users who do not need these commands will not enable the top bar, while those users (like me) who need these commands (especially the clipboard manager) would benefit greatly by enabling the top bar.

Additional enhancements might include two more modules (beside the clipboard manager, also available in the top bar):

- snippet manager (where the user may store frequently used text snippets or permanent clips, for example- names, addresses, emails, etc, something like [kboard](https://github.com/adgad/kboard) offers);
- text expander (where the user may add commands for automatic replacement (abbreviations) of text strings, for example- `.gh=GitHub`, `.gl=GitLab`, `.sf=SourceForge`, etc);

If such a top bar is implemented (which includes a clipboard manager, a snippet manager and a text expander plus the simple commands I mentioned above) in my opinion **Simple Keyboard** would become the best Android keyboard. For me addition of clipboard manager in **Simple Keyboard** is vital. If the developer is willing to also add snippet manager and text expander I may open new issues if needed.

On desktop clipboard managing, snippet managing and text automation is easy but on mobile it is not, partly because of nasty and stupid Android restrictions which prevent the use of dedicated apps.

**Comment by Cwpute** [comment]
> "," has settings, it could be added there [with a long-press], but it's so much more inconvenient than just long-press in an input field and selecting Paste.
The reason i came to love a clipboard manager in my v-keyboard is because there are many cases in which, for whatever reason, it's impossible to "just long-press in an input field". It could be the app or website that prevents it sometimes.
Also when pasting text in precise spots, trying to precisely place your big finger on the exact two-pixels that define the space between characters can be pretty darn hard !
